### ISSUE TITLE
Change detected on SHA, while SHA is reported as unchanged

### ISSUE THREAD

**Issue by rasmuslp** [fmartinou/whats-up-docker#36]
WUD sent me an email notification about new image. This looks like a false positive to me, related to the digests as interpreted by WUD:
```
Registry: hub

RegistryUrl: https://registry-1.docker.io/v2

Image: plexinc/pms-docker

Current tag: 1.22.3.4392-d7c624def

Current digest: sha256:9b5659a283844bf19b6b21fc16d9c2071cd465a254bb7df4fd21bedbc030228a

Result tag: undefined

Result digest: sha256:9b5659a283844bf19b6b21fc16d9c2071cd465a254bb7df4fd21bedbc030228a
```

Same SHA for both, and I can't figure out what image that's supposed to be.
Bonus info: `1.22.3.4392-d7c624def` doesn't look like it has changed: "Last pushed15 days ago"

Docker hub reports `sha256:2a0287ad436492c33a84a41fb143227b01f33a1ec3408de5ac084f2add38122a` for that image.
https://hub.docker.com/layers/plexinc/pms-docker/1.22.3.4392-d7c624def/images/sha256-2a0287ad436492c33a84a41fb143227b01f33a1ec3408de5ac084f2add38122a?context=explore

Which matches what docker reports on the machine running the container.
```shell
$ docker images --digests
REPOSITORY                  TAG                     DIGEST                                                                    IMAGE ID       CREATED        SIZE
plexinc/pms-docker          1.22.3.4392-d7c624def   sha256:2a0287ad436492c33a84a41fb143227b01f33a1ec3408de5ac084f2add38122a   7f33f195b496   2 weeks ago    667MB
```

I also experience this for the image `unifi-rpi` as described in #22 here https://github.com/fmartinou/whats-up-docker/issues/22#issuecomment-833697522

**Comment by rasmuslp** [comment]
I tried using [dive](https://github.com/wagoodman/dive) to check if it was the value of one of the layers, but that didn't seem to be the case.

**Comment by fmartinou** [comment]
I experienced the same behavior this morning.
Actually, the update event is not triggered because of a bad detection of a digest change but because of a bug when comparing the current tag and the new tag.
Because there is no newer tag, wud gets an 'undefined' value. And then it compares the current tag with 'undefined' and reports an available update.
I'll fix it this weekend.

**Comment by rasmuslp** [comment]
Aha, okay. Can you explain what the digest sha represents? It confuses me that it doesn't match the anything I can find on the image.

**Comment by fmartinou** [comment]
> Can you explain what the digest sha represents?

Currently it's the `Id` of the image you can get when you inspect it:
```
docker image inspect --format="{{ .Id }}" plexinc/pms-docker
```

I'll take a look to see if I can use the `RepoDigests` which is what you see on the Hub
```
docker image inspect --format="{{ .RepoDigests }}" plexinc/pms-docker
```

**Comment by rasmuslp** [comment]
This is what I get. Still different 🤷 
```
$ docker image inspect --format="{{ .Id }}" plexinc/pms-docker:1.22.3.4392-d7c624def
sha256:7f33f195b49643663d2580a0e6c8eed26e86de5d45b68b837813bbd05544a08a
```

**Comment by fmartinou** [comment]
I didn't remember but the algo to get the Digest takes `Config.Image` at first if available ([see here](https://github.com/fmartinou/whats-up-docker/blob/master/app/watchers/providers/docker/Docker.js#L170)):
```
docker image inspect --format="{{ .Config.Image }}" plexinc/pms-docker:1.22.3.4392-d7c624def
sha256:9b5659a283844bf19b6b21fc16d9c2071cd465a254bb7df4fd21bedbc030228a
```

**Comment by fmartinou** [comment]
False positive notifications should go away with version `4.0.2`.
If not, feel free to reopen the issue.

**Comment by rasmuslp** [comment]
While it is correct that a new version is available, based on the values of the tags and watch filter, it still reports the same SHA?
```
Registry: hub

RegistryUrl: https://registry-1.docker.io/v2

Image: ryansch/unifi-rpi

Current tag: 6.0.45

Current digest: sha256:ee4357041fe61908740b39601bf737447988f1473b083ebbb75399a3568f5558

Result tag: 6.1.71

Result digest: sha256:ee4357041fe61908740b39601bf737447988f1473b083ebbb75399a3568f5558
```

**Comment by fmartinou** [comment]
- ResultDigest is the digest found on the Registry **for the current tag** (`6.0.45` digest)
- ResultTag is the greater tag found on the Registry (`6.1.71`)

If one of them is different, then wud reports an available update.

ResultDigest is not the digest of the  `6.1.71` tag.
That's why you're confused, I guess.

**Comment by rasmuslp** [comment]
> * ResultDigest is the digest found on the Registry **for the current tag** (`6.0.45` digest)
> * ResultTag is the greater tag found on the Registry (`6.1.71`)
> 
> If one of them is different, then wud reports an available update.
> 
> ResultDigest is not the digest of the `6.1.71` tag.
> That's why you're confused, I guess.

Aaaah, okay. I thought ResultDigest would be the digest of what the watch expression would match (i.e. `6.1.71` in this case).

**Comment by fmartinou** [comment]
Update for amd64 is normal.
(I pushed 4.0.2 twice yesterday).

Regarding arm7, it's a bug (it was my fear).

Unfortunately, digests are messy with Docker...

I don't really now for the moment how I'll figure it out 🤔.

I keep looking for a solution...

**Comment by fmartinou** [comment]
I've opened an dedicated issue regarding the wrong digest returned for armv7: #37

**Comment by ajgon** [comment]
I think I have a similar problem (using `amd64` architecture). I'm not using emails, but this is what web interface shows me:

![digest](https://user-images.githubusercontent.com/150545/118358340-db492a00-b56d-11eb-8a76-ebbacaa894dc.png)

(By the way, this table gets out of hand, and requires a lot of horizontal scrolling - maybe an option for displaying/hiding columns in the interface? :) ). 

The digest and new digest are both the same, and the tag is the most recent. Am I missing something or is it somehow related to this bug?

**Comment by fmartinou** [comment]
What's the problem exactly?
The digest grabbed from the local image and the one found on the Hub are the same so it means that you run the latest version available for the `4.29.2`, isn't it?
Are you expecting an other result?

**Comment by ajgon** [comment]
Actually never mind, I thought that this affects `to_be_updated` field (i.e. the containers with `digest == new digest` are marked as to be updated), but it turns out it's not the case - discard my comment please :)

Out of curiosity though - why some images show new digest when the old one is the same, while other don't?
<img width="3688" alt="sha1s" src="https://user-images.githubusercontent.com/150545/118701461-44959b00-b803-11eb-9564-18c30dc265a7.png">

**Comment by fmartinou** [comment]
> why some images show new digest when the old one is the same, while other don't?

Currently there are some bugs regarding digest management.

I'm working on it (issue #37) and I hope it will solve your missing digest issue.

**Comment by fmartinou** [comment]
It's now available in version 4.1.0

**Comment by fmartinou** [comment]
Hi @ajgon 
> By the way, this table gets out of hand, and requires a lot of horizontal scrolling - maybe an option for displaying/hiding columns in the interface? :)

The UI has been totally rewritten in version `5.0.0`.
Let me know if it's moving in the right direction :)

**Comment by ajgon** [comment]
> Hi @ajgon
> 
> > By the way, this table gets out of hand, and requires a lot of horizontal scrolling - maybe an option for displaying/hiding columns in the interface? :)
> 
> The UI has been totally rewritten in version `5.0.0`.
> Let me know if it's moving in the right direction :)

I’d love to :) I’m on holidays right know but I’ll be back by the end of the week and I’ll definitely take a look 👍

**Comment by ajgon** [comment]
> Hi @ajgon
> 
> > By the way, this table gets out of hand, and requires a lot of horizontal scrolling - maybe an option for displaying/hiding columns in the interface? :)
> 
> The UI has been totally rewritten in version `5.0.0`.
> Let me know if it's moving in the right direction :)

Hey, new UI looks great, but have some comments too :) 
- Maybe it's a good idea to add `To Be Updated` to the containers filters? Since I believe usually this is the first thing, which people may look for.
- Also, is it possible to fetch all assets locally? Especially `materialicons.css` "calls" home like crazy:
  ![2021-06-27-100624_1206x285_scrot](https://user-images.githubusercontent.com/150545/123537390-ac9ea180-d71e-11eb-8a01-d5c5d07f4569.png)

Otherwise I like it, though I'm not UX designer :)

**Comment by fmartinou** [comment]
> Maybe it's a good idea to add To Be Updated to the containers filters? Since I believe usually this is the first thing, which people may look for.

There is that toggle; don't you like it ?
![image](https://user-images.githubusercontent.com/16061231/123540167-f054d380-d73d-11eb-87e0-50ccac4c5936.png)

Additionally, you can also directly navigate from the homepage to the filtered list of containers with an available update
![image](https://user-images.githubusercontent.com/16061231/123540216-3f026d80-d73e-11eb-8aae-561ee92b66df.png)

**Comment by fmartinou** [comment]
> Also, is it possible to fetch all assets locally?

By now, only fonts (including icons) are fetched remotly; is that a real issue for you?
- https://fonts.gstatic.com/s/roboto/v27/KFOmCnqEu92Fr1Mu4mxK.woff2
- https://fonts.gstatic.com/s/roboto/v27/KFOlCnqEu92Fr1MmEU9fBBc4.woff2
- https://fonts.gstatic.com/s/roboto/v27/KFOlCnqEu92Fr1MmWUlfBBc4.woff2
- https://cdn.jsdelivr.net/npm/@mdi/font@latest/fonts/materialdesignicons-webfont.woff2?v=5.9.55

=> I'll take a look if I find a way with webpack to include them in the bundle to serve them locally.

**Comment by ajgon** [comment]
> > Maybe it's a good idea to add To Be Updated to the containers filters? Since I believe usually this is the first thing, which people may look for.
> 
> There is that toggle; don't you like it ?

Somehow I missed it :facepalm: It's perfect :)

**Comment by ajgon** [comment]
> > Also, is it possible to fetch all assets locally?
> 
> By now, only fonts (including icons) are fetched remotly; is that a real issue for you?
> 
>     * https://fonts.gstatic.com/s/roboto/v27/KFOmCnqEu92Fr1Mu4mxK.woff2
> 
>     * https://fonts.gstatic.com/s/roboto/v27/KFOlCnqEu92Fr1MmEU9fBBc4.woff2
> 
>     * https://fonts.gstatic.com/s/roboto/v27/KFOlCnqEu92Fr1MmWUlfBBc4.woff2
> 
>     * https://cdn.jsdelivr.net/npm/@mdi/font@latest/fonts/materialdesignicons-webfont.woff2?v=5.9.55
> 
> 
> => I'll take a look if I find a way with webpack to include them in the bundle to serve them locally.

The problem is, it's google - despite those fonts being static assets, a tracking cookie is set :(

**Comment by fmartinou** [comment]
> The problem is, it's google - despite those fonts being static assets, a tracking cookie is set :(

All right. I'll take a look.

**Comment by fmartinou** [comment]
> Also, is it possible to fetch all assets locally?

Hi @ajgon ,
Assets are now embedded into the app so no more dependencies in Google servers ;)
[Available in version `5.1.0`](https://fmartinou.github.io/whats-up-docker/#/changelog/?id=_510)

**Comment by ajgon** [comment]
> > Also, is it possible to fetch all assets locally?
> 
> Hi @ajgon ,
> Assets are now embedded into the app so no more dependencies in Google servers ;)
> [Available in version `5.1.0`](https://fmartinou.github.io/whats-up-docker/#/changelog/?id=_510)

I checked it, and it looks gorgeous :) Thank you!
